src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,363 @@
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="chain">Thread safe class representing a chain of headers from genesis</param>
+        /// <param name="dateTimeProvider">Provider of time functions.</param>
+        /// <param name="loggerFactory">Factory to create a logger for this type.</param>
+        /// <param name="provenBlockHeaderRepository">Persistent interface of the <see cref="ProvenBlockHeader"></see> DBreeze repository.</param>
+        /// <param name="nodeLifetime">Allows consumers to perform clean-up during a graceful shutdown.</param>
+        /// <param name="nodeStats">Registers an action used to append node stats when collected.</param>
+        /// <param name="asyncLoopFactory">Factory for creating and also possibly starting application defined tasks inside async loop.</param>
+        public ProvenBlockHeaderStore(
+            ConcurrentChain chain,
+            IDateTimeProvider dateTimeProvider,
+            ILoggerFactory loggerFactory,
+            IProvenBlockHeaderRepository provenBlockHeaderRepository,
+            INodeLifetime nodeLifetime,
+            INodeStats nodeStats,
+            IAsyncLoopFactory asyncLoopFactory)
+        {
+            Guard.NotNull(chain, nameof(chain));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            Guard.NotNull(provenBlockHeaderRepository, nameof(provenBlockHeaderRepository));
+            Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+            Guard.NotNull(nodeStats, nameof(nodeStats));
+            Guard.NotNull(asyncLoopFactory, nameof(asyncLoopFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.chain = chain;
+            this.provenBlockHeaderRepository = provenBlockHeaderRepository;
+            this.nodeLifetime = nodeLifetime;
+
+            this.PendingBatch = new ConcurrentDictionary<int, ProvenBlockHeader>();
+            this.Cache = new ConcurrentDictionary<int, ProvenBlockHeader>();
+
+            this.asyncLoopFactory = asyncLoopFactory;
+
+            this.performanceCounter = new BackendPerformanceCounter(dateTimeProvider);
+            nodeStats.RegisterStats(this.AddBenchStats, StatsType.Benchmark, 400);
+        }
+
+        /// <inheritdoc />
+        public async Task InitializeAsync()
+        {
+            await this.provenBlockHeaderRepository.InitializeAsync().ConfigureAwait(false);
+
+            this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+
+            this.logger.LogDebug("Initialized ProvenBlockHader block tip at '{0}'.", this.TipHashHeight);
+
+            this.asyncLoop = this.asyncLoopFactory.Run("ProvenBlockHeaders job", token =>
+            {
+                // Save pending items.
+                this.SaveAsync().ConfigureAwait(false);
+
+                // Check and make sure the cache limit hasn't been breached.
+                this.ManangeCacheSize();
```

Comment:
By calling it here you can clean cache only once a minute which means that whatever headers we loaded on the last minute might exceed cache size. Potentially attacker can make us fill 100x the cache size if we have good enough connection speed

Reply:
Moved to `AddToPendingBatch()`.